replica/database.hh

```diff
@@ -857,10 +857,16 @@ public:
     db::replay_position set_low_replay_position_mark();
 
 private:
-    future<> snapshot(database& db, sstring name);
+    using snapshot_file_set = foreign_ptr<std::unique_ptr<std::unordered_set<sstring>>>;
+
+    future<snapshot_file_set> take_snapshot(database& db, sstring jsondir);
+    // Writes the table schema and the manifest of all files in the snapshot directory.
+    future<> finalize_snapshot(database& db, sstring jsondir, std::vector<snapshot_file_set> file_sets);
+    static future<> seal_snapshot(sstring jsondir, std::vector<snapshot_file_set> file_sets);
 
-    friend class database;
 public:
+    static future<> snapshot_on_all_shards(sharded<database>& sharded_db, const std::vector<foreign_ptr<lw_shared_ptr<table>>>& table_shards, sstring name);
+
```

Comment:
snapshot_on_all_shards() belongs better to database, but it seems you placed it here to access snapshot functions in table's private namespace.

Reply:
Right. There's a corresponding `replica::database::snapshot_table_on_all_shards` entry point.
Maybe it'd be better to declare it here private/protected and keep `friend class database`, but I don't like the database messing up with table's privates...